src/instruction/instructions/adds.ts

```diff
@@ -0,0 +1,170 @@
+import { Halfword, Word } from 'types/binary'
+
+import { Registers } from 'board/registers'
+import { IMemory } from 'board/memory/interfaces'
+import { AluResult, add } from 'board/alu'
+
+import { ILabelOffsets } from 'instruction/interfaces'
+import {
+  checkOptionCount,
+  create,
+  createLowRegisterBits,
+  getBits,
+  setBits,
+  createImmediateBits,
+  isImmediate,
+  isOptionCountValid
+} from 'instruction/opcode'
+
+import { BaseInstruction } from './base'
+
+/**
+ * Class for ADDS instruction adding 2 register values.
+ */
+export class AddsRegistersInstruction extends BaseInstruction {
+  public name: string = 'ADDS'
+  public pattern: string = '0001100XXXXXXXXX'
+  private rdPattern: string = '0001100000000XXX'
+  private rnPattern: string = '0001100000XXX000'
+  private rmPattern: string = '0001100XXX000000'
+
+  public canEncodeInstruction(name: string, options: string[]): boolean {
+    return (
+      super.canEncodeInstruction(name, options) &&
+      isOptionCountValid(options, 2, 3) &&
+      options.every((x) => !isImmediate(x))
+    )
+  }
+
+  public encodeInstruction(options: string[], labels: ILabelOffsets): Halfword {
+    checkOptionCount(options, 2, 3)
+    let opcode: Halfword = create(this.pattern)
+    opcode = setBits(opcode, this.rdPattern, createLowRegisterBits(options[0]))
+    let rnBits = createLowRegisterBits(options[options.length - 2])
+    let rmBits = createLowRegisterBits(options[options.length - 1])
+    opcode = setBits(opcode, this.rnPattern, rnBits)
+    opcode = setBits(opcode, this.rmPattern, rmBits)
+    return opcode
+  }
+
+  public executeInstruction(
+    opcode: Halfword,
+    registers: Registers,
+    memory: IMemory
+  ): void {
+    let rnRegister: number = getBits(opcode, this.rnPattern).value
+    let rmRegister: number = getBits(opcode, this.rmPattern).value
+    let rdRegister: number = getBits(opcode, this.rdPattern).value
+
+    let rnRegisterContent: Word = registers.readRegister(rnRegister)
+    let rmRegisterContent: Word = registers.readRegister(rmRegister)
+
+    let aluResult: AluResult = add(rnRegisterContent, rmRegisterContent)
+
+    registers.writeRegister(rdRegister, aluResult.result)
+    registers.setFlags(aluResult.flags)
+  }
+}
+
+/**
+ * Class for ADDS instruction adding a register value and an immediate value.
+ */
+export class AddsImmediate3Instruction extends BaseInstruction {
+  public name: string = 'ADDS'
+  public pattern: string = '0001110XXXXXXXXX'
+  private rdPattern: string = '0001110000000XXX'
+  private rnPattern: string = '0001110000XXX000'
+  private immPattern: string = '0001110XXX000000'
+  private expectedOptionsCount: number = 3
+
+  public canEncodeInstruction(name: string, options: string[]): boolean {
+    return (
+      super.canEncodeInstruction(name, options) &&
+      isOptionCountValid(options, this.expectedOptionsCount) &&
+      isImmediate(options[2])
```

Comment:
additional check if option[0] and option[1] differ otherwise AddsImmediate8Instruction is never used when rdn is provided twice

Reply:
Good catch 👍